### Analyte Classification using Logistic Regression
#### Edgar Acuna 
#### July 2020

In [1]:
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report, confusion_matrix, f1_score, precision_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
%matplotlib inline

#### Loading the data

In [2]:
df1=pd.read_csv("c://onr2020/NRLset1_part1.csv",header=None)
df2=pd.read_csv("c://onr2020/NRLset1_part2.csv",header=None)
df3=pd.read_csv("c://onr2020/NRLset1_part3.csv",header=None)
df4=pd.read_csv("c://onr2020/NRLset1_part4.csv",header=None)
df5=pd.read_csv("c://onr2020/NRLset1_part5.csv",header=None)
df6=pd.read_csv("c://onr2020/NRLset1_part6.csv",header=None)
df7=pd.read_csv("c://onr2020/NRLset1_part7.csv",header=None)
df8=pd.read_csv("c://onr2020/NRLset1_part8.csv",header=None)
y=pd.read_csv("c://onr2020/labels.csv",header=None)
ys=pd.read_csv("c://onr2020/substrateIDs.csv",header=None)

In [3]:
dfset1=pd.concat([df1,df2,df3,df4,df5,df6,df7,df8],ignore_index=True)
dfset1.head()

0         1         2         3         4         5         6     \
0  0.041418  0.041621  0.042198  0.042688  0.042924  0.042274  0.042542   
1  0.694710  0.695840  0.697180  0.706530  0.703970  0.709130  0.711480   
2  0.048978  0.048432  0.047685  0.047086  0.046811  0.046752  0.046624   
3  0.039762  0.039495  0.038982  0.038339  0.037769  0.037301  0.036799   
4  0.022387  0.022508  0.022091  0.023054  0.023010  0.022740  0.023889   

       7         8         9     ...      1691      1692      1693      1694  \
0  0.042663  0.042715  0.042664  ...  0.053219  0.053131  0.053091  0.053140   
1  0.711330  0.712370  0.712980  ...  0.754260  0.751040  0.753630  0.754490   
2  0.046443  0.046241  0.045999  ...  0.260320  0.259590  0.258910  0.258210   
3  0.036316  0.035921  0.035612  ...  0.257600  0.256890  0.256160  0.255370   
4  0.023936  0.023464  0.024810  ...  0.025036  0.024307  0.024636  0.022298   

       1695      1696      1697      1698      1699      1700  
0  0.053249  0.053325  0.053364  0.053356  0.053500  0.055986  
1  0.752980  0.755760  0.752910  0.755320  0.753030  0.753230  
2  0.257440  0.256580  0.255700  0.254800  0.254160  0.257110  
3  0.254540  0.253720  0.252880  0.251970  0.251070  0.250190  
4  0.023536  0.025714  0.025306  0.025062  0.023609  0.023901  

[5 rows x 1701 columns]

In [4]:
dfset2=dfset1.copy()
dfset2['Analyte']=y
dfset2['substrate']=ys
dfsub1=dfset2[dfset2['substrate']==1]
dfsub2=dfset2[dfset2['substrate']==2]
dfsub3=dfset2[dfset2['substrate']==3]
dfsub4=dfset2[dfset2['substrate']==4]
dfsub5=dfset2[dfset2['substrate']==5]
dfsub6=dfset2[dfset2['substrate']==6]
dfsub7=dfset2[dfset2['substrate']==7]
dfsub8=dfset2[dfset2['substrate']==8]
dfsub9=dfset2[dfset2['substrate']==9]
dfsubset0=pd.concat([dfsub3,dfsub4,dfsub9],ignore_index=True)

In [5]:
dfsub61 = dfsub6.drop(dfsub6[(dfsub6['Analyte']==13) | (dfsub6['Analyte']==24) | (dfsub6['Analyte']==34)].index)
dfsub91 = dfsub9.drop(dfsub9[(dfsub9['Analyte']==13) | (dfsub9['Analyte']==24) | (dfsub9['Analyte']==34)].index)
dfsub91.head()

0        1        2        3        4        5        6        7  \
6   0.23532  0.23367  0.23625  0.24053  0.23726  0.23638  0.23661  0.23843   
11  0.62481  0.62296  0.61857  0.61310  0.60450  0.60094  0.59381  0.58911   
16  0.26241  0.26114  0.26299  0.26278  0.26165  0.25819  0.25645  0.25475   
24  0.30203  0.30009  0.31935  0.35520  0.35301  0.34536  0.34466  0.33122   
26  0.21222  0.15344  0.17677  0.17939  0.19517  0.17923  0.16505  0.16720   

          8        9  ...     1693     1694     1695     1696     1697  \
6   0.23874  0.23832  ...  0.22309  0.22392  0.22221  0.22300  0.22181   
11  0.59014  0.58750  ...  0.47459  0.47420  0.47397  0.47378  0.47421   
16  0.25381  0.24845  ...  0.38369  0.38390  0.38142  0.38227  0.38204   
24  0.31698  0.32534  ...  0.40309  0.39981  0.40789  0.41994  0.40553   
26  0.17508  0.15312  ...  0.37571  0.37403  0.39967  0.38927  0.38026   

       1698     1699     1700  Analyte  substrate  
6   0.22201  0.22098  0.22148        5          9  
11  0.47082  0.47286  0.47047       36          9  
16  0.38286  0.38272  0.38236       10          9  
24  0.42968  0.40836  0.44263       20          9  
26  0.39495  0.37657  0.39560        3          9  

[5 rows x 1703 columns]

### 6. Applying knn with k=3  and two classes: Analyte j (j=1,,..40) versus other analytes (class=0)). 

In [34]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver="newton-cg",max_iter=5000)
y=dfsub4['Analyte']
X=dfsub4.iloc[:,0:1701]
model.fit(X,y)
model.score(X,y)
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.1,random_state=0)
# Applying again the k-nn classifier with k=3
#neigh = KNeighborsClassifier(n_neighbors=kVals[i])
model=LogisticRegression(max_iter=5000) 
model.fit(X_train,y_train)
#Calculating  metrics of prediction
predictions = model.predict(X_test)
print(predictions)
print("F1-score is=",f1_score(y_test,predictions, average="weighted"))
print("precision=",precision_score(y_test,predictions,average="weighted"))
print("The accuracy is=",model.score(X_test,y_test))

[21  8 23 17 25  2 21 27 27 13  5 38  9 31 14 19  5 40  9 16 32 16 26 10
  1 22  1 29 28 19 14 31 19 40 25 22 39 17 22 40 40 14  1 15 20  1 14 12
 37 21 21  1 27 34 38 33  1 38 22 26 17  7 19 10 20 32 30 30 26 10  2 10
 21 14 35 31 25 19  9 11 39 25 26  8  4 39 14 15 17 11 39 39 22 31 23 22
 11 20  5 33 34 14 21 31 24 29 14 26 28 35  8 33 31 32 27  6 39 27 21 20
 16 23 34 39 34  7 32 15 12 29 30 34 24 15 31 33 12 40 40  3  4  2 38 36
 29  1 24 35  7 40 20 25  4 21  2 17 26 16 17 32 21 29 36 32 33 33 35  4
  3  4 25  9 28 27 34  5 34 30 36 32 11 14 10 34 15 10 30 29 17 25  9  5
  2 25 29  7 14  4  6 14]
F1-score is= 0.9654399766899766
precision= 0.9777678571428571
The accuracy is= 0.965


In [1]:
#Only the first twenty analytes are analyzed
labels=list(range(1,41))

In [72]:
for j in labels:
    yclass=dfsub1['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub1.iloc[:,0:1071],yclass,test_size=0.2,random_state=0)
    scaler = StandardScaler()
    scaler.fit(X_train)
    X_train= scaler.transform(X_train)
    X_test = scaler.transform(X_test)
    clf=svm.SVC()
    clf.fit(X_train, yclass_train) 
    predictions = clf.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, clf.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=0.980, F1-score=0.200, Precision=1.000
Metrics for predicting analyte 2 : accuracy=0.983, F1-score=0.667, Precision=1.000
Metrics for predicting analyte 3 : accuracy=0.980, F1-score=0.200, Precision=1.000
Metrics for predicting analyte 4 : accuracy=0.978, F1-score=0.308, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.980, F1-score=0.636, Precision=1.000
Metrics for predicting analyte 6 : accuracy=0.973, F1-score=0.267, Precision=1.000
Metrics for predicting analyte 7 : accuracy=0.988, F1-score=0.737, Precision=1.000
Metrics for predicting analyte 8 : accuracy=0.975, F1-score=0.375, Precision=1.000
Metrics for predicting analyte 9 : accuracy=0.993, F1-score=0.769, Precision=1.000
Metrics for predicting analyte 10 : accuracy=0.985, F1-score=0.571, Precision=1.000
Metrics for predicting analyte 11 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 12 : accuracy=0.978, F1-score=0.000, Precision=0.000
M

In [87]:
for j in labels:
    yclass=dfsub6['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub6.iloc[:,0:1071],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=0.995, F1-score=0.909, Precision=1.000
Metrics for predicting analyte 2 : accuracy=0.995, F1-score=0.889, Precision=1.000
Metrics for predicting analyte 3 : accuracy=0.995, F1-score=0.800, Precision=1.000
Metrics for predicting analyte 4 : accuracy=0.995, F1-score=0.667, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.995, F1-score=0.857, Precision=1.000
Metrics for predicting analyte 6 : accuracy=0.985, F1-score=0.571, Precision=0.500
Metrics for predicting analyte 7 : accuracy=0.995, F1-score=0.800, Precision=1.000
Metrics for predicting analyte 8 : accuracy=0.995, F1-score=0.923, Precision=1.000
Metrics for predicting analyte 9 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 12 : accuracy=1.000, F1-score=1.000, Precision=1.000
M

In [94]:
for j in labels:
    yclass=dfsub9['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub9.iloc[:,0:1071],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 2 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 3 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 4 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.995, F1-score=0.857, Precision=1.000
Metrics for predicting analyte 6 : accuracy=0.995, F1-score=0.923, Precision=1.000
Metrics for predicting analyte 7 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 8 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 9 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 12 : accuracy=1.000, F1-score=1.000, Precision=1.000
M

In [95]:
labels1=[1,2,3,4,5,6,7,8,9,10,11,12,14,15,16,17,18,19,20,21,22,23,25,26,27,28,29,30,31,32,33,35,36,37,38,39,40]
for j in labels1:
    yclass=dfsub91['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub91.iloc[:,0:1071],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 2 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 3 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 4 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.995, F1-score=0.889, Precision=1.000
Metrics for predicting analyte 6 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 7 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 8 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 9 : accuracy=0.995, F1-score=0.933, Precision=1.000
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 12 : accuracy=1.000, F1-score=1.000, Precision=1.000
M

#### The F1-score is above .90 for six  analytes and,  between .80-.90 for 22 analytes .  The analytes 13(0.652),18(0.618),24(0.602) and 34(0.601) have the lowest F1-score. Also, they have the lowest auc.

#### The prediction results improve. Now, the F1-score is above .90 for twenty-eight analytes and, between .80-.90 for nine analytes . The analytes 13(0.757), 24(0.602) and 34(0.760) have the lowest F1-score. Also, they have the lowest auc.